### DISCUSSION TITLE
Question: Are there currently any file picker options available when running a WinUI3 app in PartialTrust mode?

### DISCUSSION THREAD

**Discussion by karter16** [microsoft/WindowsAppSDK#3447] [Q&A]
Are there currently any file picker options available when running a WinUI3 app in PartialTrust mode?

I have an existing UWP app which I have chosen to migrate to a WinUI3-based app. I am building and packaging the app in PartialTrust mode for the following reasons:

1: Having been a UWP app to begin with the app was already running inside an AppContainer, etc. 
2: I want to continue to be able to offer the app on the MS Store without having to request FullTrust for the app, which is unnecessary. The app is for use on PCs that are used during live productions and as a result my customers are fairly suspicious of anything that requires unfettered access to their devices, I'd therefore much prefer to not have to require FullTrust.

What I'm looking to understand is whether there are any file-picker-like options available to me at this point in time. I've done a fair amount of reading and investigation and as I understand it I'm not able to go the file picker route, as using a file picker requires giving it an HWND and WinRT.Interop.InitializeWithWindow can't be run in PartialTrust mode, so from what I can tell that option is off the table for me.

Are there any other suggestions/options that I could look at at this point in time (short of building my own filepicker type of thing)? Any suggestions/leads would be appreciated.

Note: I do realise that I have brought this on myself, my fallback option is to keep on as a UWP app, I just like to keep up to speed with the latest developments and options, hence wanting to explore this.

Thanks,

Matt

**Comment by DarranRowe** [comment]
What makes you think the file pickers can't be used in partial trust mode? All partial trust does is put the application into a lightweight app container, it is still running as a desktop application.

**Reply by karter16** [reply]
The issue is that for a WinUI3 app, because there is potentially more than 1 window (unlike a UWP app) you need to pass the HWND to the filepicker when initialising it. e.g.:

` WinRT.Interop.InitializeWithWindow.Initialize(picker, App.WindowHandle);`

This call results in an exception when running in PartialTrust mode:

`System.UnauthorizedAccessException: 'Access is denied. (0x80070005 (E_ACCESSDENIED))'`

**Reply by DarranRowe** [reply]
I have been messing around with this in a pure Windows API C++ application. To be honest, I would seriously put this into the "It is a Windows bug that Microsoft must fix" category. This is also easier than rolling your own because there is the whole issue of Windows.Storage.Pickers use a broker to do its work in an application in an app container. From the documentation, brokers require sideloading, so that is not useful for store applications.
While I was messing around, I did try to use the shell's file pickers, but they gave a lot of wonderful access denied errors since they were in process and therefore couldn't access anything but AppData.
The documentation isn't very good here anyway, so it is hard to know if this is intentional or not, but since the pickers are the only ones that currently adhere to the UWP security model which is in play even in a light weight app container, then I find it hard to believe that this is intentional.
What is even more curious is that there doesn't seem to be a single report in the feedback hub. So is this a case of everyone saying "it doesn't work" but then nobody reporting it as a bug?

--Edit--

Well, I did create a bug report for this on the feedback hub. You can find the link [here](https://aka.ms/AAjow6x). Try to get as many people to upvote this as possible to make Microsoft aware that you want this fixed.

**Reply by karter16** [reply]
Hi @DarranRowe - apologies for the delay in getting back to you, I was unwell over the weekend.
Thanks very much for the time you've put into looking at this. Yes it was the fact that there wasn't any bug reports about it that made me wonder what I was missing!

Thanks very much for creating the bug report, I'll head across and upvote there now as it would be great to have a solution to this - seems like a bit of a gap at the moment!

Thanks again for your help :-)

**Reply by jaigak** [reply]
The problem here is that the file picker under partial trust assumes there's a CoreWindow on the current thread, which is obviously not true in WinUI 3 apps.

**Reply by karter16** [reply]
> The problem here is that the file picker under partial trust assumes there's a CoreWindow on the current thread, which is obviously not true in WinUI 3 apps.

Yes that's right, CoreWindow doesn't exist in WinUI 3 land, and you obviously can't pass an HWND in PartialTrust mode, hence anyone wanting to use a filepicker in WinUI 3 PartialTrust mode is stuck at present.

**Reply by jaigak** [reply]
For now, you can enable broad file system access and create your own file picker. Obviously not a good idea.

**Reply by Rybasum** [reply]
https://github.com/microsoft/microsoft-ui-xaml/issues/4100#issue-801059805
This is an old issue and Microsoft seems not to care. File pickers is something most apps do. Hence, I infer they don't want us to put non-UWP apps in AppContainer for some reason.

**Reply by DarranRowe** [reply]
@Rybasum
There is something else people could do to help.
Since the pickers and app containers are a Windows component, this should be reported through the Feedback Hub. I did open a bug report for this over a year ago and nobody has been upvoting it. I also can't find any other bug reports on the Feedback Hub related to the pickers. So, when the only bug report on the Feedback Hub related to the pickers has 2 upvotes, how important does that make this issue seem?

**Reply by ejhg** [reply]
What a wonderful situation.  I couldn't get the workaround in https://github.com/microsoft/microsoft-ui-xaml/issues/4100#issue-801059805 to work.

I tried the bug report link but was unable to open it.

**Reply by DarranRowe** [reply]
I think by now adding more support for the WAS team to implement a new set of pickers is the way to go.

Currently #8 is being used as the feature proposal to implement a new set of storage functionality. As always, replying/reacting to this feature proposal will help.
In the 1.7 plans post, I [also added a suggestion](https://github.com/microsoft/WindowsAppSDK/discussions/4710#discussioncomment-10617202) to implement these new pickers. Again, the more upvotes these get the better. Basically, community demand is required here.

**Reply by ejhg** [reply]
upvoted those 2 links.  thanks @DarranRowe !

**Reply by ejhg** [reply]
For now, I'll proceed by enabling full trust mode just to be able to open and save files in an app that would otherwise not need it.  It's wild that doing the "right thing" (ie, least privilege principle) is so hard.

(Windows, vulnerable by design.)
